### DISCUSSION TITLE
Retrieving submodule from module declared with partial

### DISCUSSION THREAD

**Discussion by rolandgvc** [google/flax#665] [Q&A] [answered]
Given the following example:
```python
class Net(nn.Module):
  ...
  
  def setup(self):
      self.backbone = Backbone(...)
      self.classifier = nn.Dense(...)

    def __call__(self, inputs):
      x = self.backbone(inputs)
      x = self.classifier(x)
      return x

def create_model(rng, private_params, **kwargs):
  model = functools.partial(Net, private_params, **kwargs)
  init_batch = jnp.ones((1, 224, 224, 3), jnp.float32)
  params = Net(**kwargs).init(rng, init_batch)
  return model, params

def net_50(rng, **kwargs):
  return _create_model(rng, private_params, **kwargs)

model, params = net_50(RNG, train=False)
```

How would it be possible to extract the backbone submodule from `model` while still using `functools.partial` to hide the private params from the user?

**Comment by avital** [comment, marked as answer]
Hi Roland, can you return an instance of `Net` instead a function that creates it? We need to document this better but just calling `Dense(3)` doesn't actually create any variables, it's just a skeleton "module template" (what we call an "unbound module" in the code)?

Though I think at the moment you still won't have access to `Net().backbone` because `setup` is only called when the module is [bound](https://flax.readthedocs.io/en/latest/flax.linen.html#flax.linen.Module.setup)... This actually may be good to improve.

Can you say more about your full use-case for accessing `net.backbone` in your code?

**Reply by rolandgvc** [reply]
Hi Avital, thanks for your reply. I'm looking to extract the backbone module of a CNN. For a concrete example, take a look at this implementation of ResNet https://github.com/rolandgvc/flaxvision/blob/ff0aca932938857f959ae855579297b2e9b9f50a/flaxvision/models/resnet.py#L205

For implementing a segmentation model (or doing transfer learning in general) I want to extract the `backbone` module and attach a new classifier to it. Initially I thought access to modules defined in `setup` would be similar to PyTorch's, but maybe that's not the case. What would be your approach for achieving this? Thanks!

**Reply by rolandgvc** [reply]
Something else I didn't mention in the example, one of the `private_params` is `pretrained`, which in my current implementation will pass the pretrained weights instead of initializing them from a dummy batch. I assume this behavior must change as well if I'm interested in extracting the backbone.

**Reply by avital** [reply]
I think what you're asking for is something like this?

```
class SegmentationModel(nn.Module):
  backbone: Module

  def setup(self):
    # here you can access `self.backbone`

  def __call__(self, x):
    # here you can also access `self.backbone`

SegmentationModel(backbone=flaxvision.ResNet().backbone, ...)
```

I think indeed this doesn't work at the moment (because `setup` is only called when a module is bound, not when it's constructed outside of a module like this). Let me think more about this and propose a solution.

(I didn't understand your comment about `private_params`. What do you end up doing with model afterwards? Don't you always call `model()(...)`? If so you should be able to just return the module instance instead of a partial)

**Reply by avital** [reply]
I actually now think my proposal doesn't work. Let me think a bit more about this and I'll propose a working pattern.

**Reply by rolandgvc** [reply]
Thanks Avital! Here is an example of what the `private_params` could be https://github.com/rolandgvc/flaxvision/blob/ff0aca932938857f959ae855579297b2e9b9f50a/flaxvision/models/resnet.py#L257
I wouldn't want the user to be passing the block type or layer config at initialization, but simply be able to call:
```python
model, params = flaxvision.models.resnet50(RNG)
out = model(train=False).apply(params, input, mutable=False)
```
Did I get your message correctly?

**Reply by avital** [reply]
In https://github.com/rolandgvc/flaxvision/blob/ff0aca932938857f959ae855579297b2e9b9f50a/flaxvision/models/resnet.py#L244, I think you can just return `ResNet(block=block, layers=layers, **kwargs)` instead of using `partial`. Then you have access to the `ResNet` object.

One tricky thing with the current API is that `setup` is only called when a module is "bound" -- e.g. during `init` or `apply` (see https://flax.readthedocs.io/en/latest/flax.linen.html#flax.linen.Module.setup for more details). Your example shows that `setup` is tricky because it's not identical to the semantics of `__init__`. (There's a possible API change we could make to call `setup` before the module is bound, and only bind submodules and parameters later but there are many questions to think through)  This use-case is indeed evidence that we should think a bit more deeply.

Concretely, in your case, even if you have access to a `ResNet` instance, you won't be able to access `resnet.backbone`. 

In the meanwhile, an approach that I think will work is to introduce `make_backbone()` into a method (see code below). 

```
class ResNet(nn.Module):
  ...

  def make_backbone(self):
    return Backbone(self.block, self.layers, self.num_classes, self.groups,
                             self.width_per_group, self.replace_stride_with_dilation,
                             self.train, self.dtype)

  def setup(self):
    self.backbone = self.make_backbone()
    self.classifier = nn.Dense(self.num_classes, dtype=self.dtype)

  ...
```

Then you can access `resnet.make_backbone()` from anywhere you want, e.g.

```
class UNet(nn.Module):
  def setup(self):
    self.backbone = ResNet().make_backbone()
    ...
```

Please let me know if this helps, and we will think more about this "design test".

**Reply by rolandgvc** [reply]
Awesome! Will try it out 👍

**Reply by rolandgvc** [reply]
Sorry I haven't gotten to try your suggestion until now. You mentioned that I can return the `ResNet` instance directly. In this case, how can I still be switching the `deterministic` parameter of batch norm and dropout at calling? Is it still possible?
```python
model, params = flaxvision.models.resnet50(RNG, train=False)
out = model.apply(params, input, mutable=False) # can train still be passed here?
```

**Reply by rolandgvc** [reply]
Calling `resnet.make_backbone()` results in this error:
``` 
File "/Users/rolandgavrilescu/miniconda3/lib/python3.7/site-packages/flax/linen/module.py", line 148, in wrapped_module_method
    raise ValueError("Can't call methods on orphaned modules")
ValueError: Can't call methods on orphaned modules
```

**Reply by avital** [reply]
That last error should have been fixed by https://github.com/google/flax/pull/653. (It's on latest GitHub version, I haven't yet released a new PyPI version). Can you try installing Flax from GitHub?

Re `determinstic` -- I think maybe better if you turn deterministic into an argument to your `ResNet.__call__`. Then you're using the same module instance. (I admit we used the wrong pattern in our imagenet example a while back but now we do it this was as well: https://github.com/google/flax/blob/master/linen_examples/imagenet/models.py#L92)

**Reply by rolandgvc** [reply]
Thanks! Getting this error now:
```
File "/Users/rolandgavrilescu/miniconda3/lib/python3.7/site-packages/flax/linen/module.py", line 476, in __post_init__
    'Submodules must be defined in `setup()` or in a method wrapped '
ValueError: Submodules must be defined in `setup()` or in a method wrapped in `@compact`
```
Am I doing this correctly?
```python
class ResNet(nn.Module):
  block: Any
  layers: Any
  num_classes: int = 1000
  groups: int = 1
  width_per_group: int = 64
  replace_stride_with_dilation: Any = None
  dtype: Any = jnp.float32

  def make_backbone(self):
    return Backbone(self.block, self.layers, self.num_classes, self.groups,
                    self.width_per_group, self.replace_stride_with_dilation,
                    self.dtype)

  def setup(self):
    self.backbone = self.make_backbone()
    self.classifier = nn.Dense(self.num_classes, dtype=self.dtype)

  def __call__(self, inputs, train: bool = False):
    x = self.backbone(inputs, train)
    x = self.classifier(x)

    return x
```

**Reply by avital** [reply]
You're right @rolandgvc -- my proposed workaround... doesn't work. Fundamentally the solution to reducing this layer of magic is probably going look like #686. (in which case both your original `setup` version and my proposed version with methods would work), and in general `setup` works a lot more like a normal constructor.

For now, here's a workaround that I verified works. It's ugly, I'll admit. 

```
import jax
from flax import linen as nn
from flax.linen import Dense


class Backbone(nn.Module):
  dtype: Any = jnp.float32

  @nn.compact
  def __call__(self, inputs):
    return Dense(10, dtype=self.dtype)(inputs)


class ResNet(nn.Module):
  num_classes: int = 10
  dtype: Any = jnp.float32

  @staticmethod
  def make_backbone(self):
    return Backbone(self.dtype)

  def setup(self):
    self.backbone = ResNet.make_backbone(self)
    self.classifier = nn.Dense(self.num_classes, dtype=self.dtype)

  def __call__(self, inputs, train: bool = False):
    x = self.backbone(inputs)
    x = self.classifier(x)

    return x


r = ResNet()
backbone = ResNet.make_backbone(r)

init_vars = r.init(jax.random.PRNGKey(8), jnp.ones((2, 2)))

backbone.apply({'params': init_vars['params']['backbone']}, jnp.ones((2, 2)))
```

**Reply by rolandgvc** [reply]
Thanks for the help Avital. However, I still can't get it to work with my use case. Am I doing it wrong?
```python
import resnet
from fcn import FCNHead

class SegmentationModel(nn.Module):
    backbone: Any
    classifier: Any

    @nn.compact
    def __call__(self, inputs, train: bool = False):
        x = self.backbone(inputs, train=train)
        x = self.classifier(x, train=train)
        return x


def _make_model(rng, head_name, backbone_name, num_classes, **kwargs):
    segm_heads = { 'fcn': FCNHead }

    resnet_model, _ = resnet.__dict__[backbone_name](rng)
    backbone = resnet.ResNet.make_backbone(resnet_model)
    classifier = segm_heads[head_name](num_classes)

    return SegmentationModel(backbone, classifier, **kwargs)


def _load_model(rng, arch_type, backbone, pretrained, num_classes, **kwargs):
    model = _make_model(rng, arch_type, backbone, num_classes, **kwargs)
    if pretrained:
        flax_params = load_flax_params()
    else:
        init_batch = jnp.ones((1, 224, 224, 3), jnp.float32)
        flax_params = model(**kwargs).init(rng, init_batch) # ValueError: Can't call compact methods on unbound modules

    return model, flax_params
```
The resnet implementation is at https://github.com/rolandgvc/flaxvision/blob/linen/flaxvision/models/resnet.py

**Reply by avital** [reply]
Hi Roland, it would really help to have a Colab that I can run directly to see the problem, it's a bit hard to combine this code snippet with the latest versions of FlaxVision and Flax. (Generally that's going to always be helpful, it takes quite some time to context switch to look at a user report so the easier it is to debug, the faster we can respond)

**Reply by avital** [reply]
(I tried doing `!pip install -q git+https://github.com/rolandgvc/flaxvision/tree/linen/` in a Colab but then I got: `ERROR: flaxvision 0.0.5 has requirement flax==0.2.0, but you'll have flax 0.3.0 which is incompatible.`)

**Reply by rolandgvc** [reply]
Right, sorry for that. I'm putting a colab together now. Thanks for taking the time!

**Reply by rolandgvc** [reply]
Here it is https://colab.research.google.com/drive/1KAQspBG3BEJ2PgNNtd1n63kgf78PPpZU?usp=sharing

**Reply by avital** [reply]
Thanks Roland! I revised your example and got it working (I had to change a few other things as well -- you didn't need `@compact` at all around `SegmentationModel.__call__`, and your call to init should have `model.init(rng, init_batch)
`): https://colab.research.google.com/drive/1QlwImBjxZxMFkM50qR9FzJ4o-QW51Waa#scrollTo=J4So4-Sf8oUN

The important part of the change is this pattern:

```py
class SegmentationModel(nn.Module):
    backbone_fn: Any
    classifier_fn: Any

    def setup(self):
        self.backbone = self.backbone_fn()
        self.classifier = self.classifier_fn()

    def __call__(self, inputs, train: bool = False):
        x = self.backbone(inputs, train=train)
        x = self.classifier(x, train=train)
        return x

def _make_model(rng, train, **kwargs):
    resnet_model, _ = resnet50(rng)
    backbone_fn = lambda: ResNet.make_backbone(resnet_model)
    classifier_fn = lambda: FCNHead(10, train=train)

    return SegmentationModel(backbone_fn, classifier_fn, **kwargs)
```

That is -- pass in a function that makes a module as an attribute into a larger module, and then assign it during setup. The assignment during setup is what attaches it to the parent module, so there must be explicit assignment. Simply passing in attributes doesn't go through that (intentionally - because you may use a submodule somewhere else in the module tree and then it's incorrect to bind it twice).

We really need to improve this user experience, and I hope to spend some more time on #686, thanks for this motivating example!

Please let me know if this last suggestion seems to work for you. If you hit any more issues, please continue to post here (or as a new discussion, as appropriate).

Thanks for slogging with us -- this is extremely useful feedback and we need this to keep improving Flax!

**Reply by rolandgvc** [reply]
Awesome!! Thanks @avital and the team for always being active and helpful in these discussions! I will try the example in a bit. In case I still struggle with it, I will create a new discussion and link it here (since this one deviated a bit from the original topic). Many thanks!
